### PR TITLE
Generalize cube extract

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request generalizes cubeviz's spectral extraction so that it can be re-used by lcviz (photometric extraction of target pixel files) and eventually rampviz.  This is milestoned for the 4.0 release, which already contains breaking changes to spectral extraction.

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py
Comment (rosteen) [inline jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:200]: This would be a good time to update this to be more informative/correspond better with what is actually being checked. I'd change it to mention that the plugin is expecting a _three dimensional_ dataset to be loaded.

**Review by rosteen — APPROVED** [review]
One minor suggestion about some disabled message text, but other than that I think this looks good - less complicated that the diff suggested, mostly just moving things around and renaming.

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py
Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:200]: Good point, I'll make that change.

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py
Comment (bmorris3) [inline jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:193]:
```suggestion
            if data.ndim == 3:
```

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py
Comment (bmorris3) [inline jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:222]: Is `ext` for "extraction", rather than its usual "extension" as abbreviated elsewhere in jdaviz? This would require changes throughout the PR, but could we use a less degenerate abbreviation? Maybe `extr`?

**Review by bmorris3 — COMMENTED** [review]
Some superficial comments for readability.

File: jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py
Comment (bmorris3) [reply jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:222]: (While I'm aware that I can be a pedant,) more descriptive names are always better.

**Comment by kecnry** [comment]
@bmorris3 / @rosteen - added a few more commits since your last reviews if you want to re-review (should be able to filter the diff to just the last ~6 commits)

@bmorris3 - addressed your comments.  I haven't touched renaming background or wavelength-dependent related methods and attributes that don't need to be used for lcviz (`wavelength_dependent`, `bg_wavelength_dependent`, `bg_spec_per_spaxel`, `bg_spec_add_results`, `extract_bg_spectrum`) - but will rampviz need access to these and maybe we should generalize them (either here or as a separate effort before they get released in 4.0)?

**Comment by rosteen** [comment]
I seem to be getting some looping behavior now, could this be due to removing `skip_if_no_updates_since_last_active`? It did stop eventually but took a lot longer than I would expect. This was after selecting a spatial subset for a sum extraction.

https://github.com/user-attachments/assets/5b815992-0fce-436d-b63e-f9ec4c8d5a07

**Comment by kecnry** [comment]
drats, I wasn't seeing that but that is most likely the cause.  I guess I'll have to split that method after all.

File: jdaviz/core/template_mixin.py
Comment (bmorris3) [inline jdaviz/core/template_mixin.py:2448]: I was kind of mystified that this line works. I checked that it indeed works, but it feels like numerology? I've put in an equivalent but maybe less coincidental implementation below for your consideration.

```python
spatial_axes = [(0, 1), (1, 2), (0, 2)]
answers = [2, 0, 1]
for input, output in zip(spatial_axes, answers):

    assert 3 - sum(input) == output
    
    # suggested alternative:
    assert set([0, 1, 2]).difference(input).pop() == output
```

```suggestion
        slice_axis = set([0, 1, 2]).difference(input).pop()
```

**Review by bmorris3 — APPROVED** [review]
Tested it out and it looks great. Thanks!

**Comment by rosteen** [comment]
In addition to the preview not updating properly as reported offline, I'm a little worried that the extraction calculation is not correct either when a spatial subset is selected. In the screenshot below you can see that I selected a spatial subset that covers basically the whole spatial extent of the cube, but the extracted sum is significantly lower than the sum for the cube with no spatial subset.

Screenshot 2024-07-15 at 10 51 41 AM

**Comment by kecnry** [comment]
@rosteen - is this different than on main?

**Comment by rosteen** [comment]
Hmm, turns out it is present on main. Here you can see that even a subset covering the full spatial extent ends up with a smaller sum:

Screenshot 2024-07-15 at 11 12 49 AM

**Comment by rosteen** [comment]
See https://github.com/spacetelescope/jdaviz/issues/3086

File: jdaviz/core/template_mixin.py
Comment (kecnry) [reply jdaviz/core/template_mixin.py:2448]: I can see how that's a little more obvious to read (although I do think the sum logic makes sense and isn't just a coincidence - it just only works when you want to find the single missing number from a set), but probably less performant?

**Comment by kecnry** [comment]
@rosteen - can you take another look now and see if the live-previews now act as you'd expect?  Some of this buggy behavior was on main previously, but I do think this is a good chance to fix them, so would like to get it right if possible.

**Comment by kecnry** [comment]
> I'm still a little surprised that moving/resizing a subset doesn't update the preview, but I think that's just because it's hitting the >1 second update time that triggers the warning and requires a manual click to update.

I can try to implement this (I think its the same on main - but agreed that it isn't correct).

> Related to that (but probably not related directly to this PR), I'm a little surprised that if you click "Disable Previews" when they're taking too long, there doesn't seem to be a way to ever turn them back on from the UI. Maybe I'm missing something, and this could be a follow up anyway.

They can be re-enabled from the "settings" section at the top of the plugin.  This is mentioned in the tooltip text, but maybe should be more prominent somewhere.  That feels like a separate effort to me that might need more input, so feel free to create a follow-up ticket if you have any ideas/suggestions.

**Comment by kecnry** [comment]
>> I'm still a little surprised that moving/resizing a subset doesn't update the preview, but I think that's just because it's hitting the >1 second update time that triggers the warning and requires a manual click to update.

> I can try to implement this (I think its the same on main - but agreed that it isn't correct).

This is proving more difficult than I had hoped.  On main, it seems that the preview is just dropped when the subset is updated, so I don't think this is a reversion, but because the _visibility_ of the preview has been fixed here, this does expose this "bug".  Would you be ok with this as a follow-up ticket or do you think that because it exposes this bug more clearly (by fixing the visibility bug) that the bug should block this PR?

**Comment by kecnry** [comment]
I think (again, hopefully for real this time) that I have the live-previews updating correctly for a change to the selected subset.  See the last ~5 commits for the diff since your previous reviews.  @rosteen @bmorris3 - please take a quick look/test and let me know if you have any comments or if your approvals still stand.

**Comment by rosteen** [comment]
I think it looks ok, my approval can stand.

File: jdaviz/core/template_mixin.py
Comment (bmorris3) [reply jdaviz/core/template_mixin.py:2458]: I've tested it, and of course it is less performant – but it's the difference between 40 nanoseconds for the original and 150 ns for my suggestion. I'm not sure anyone will notice. 😉

**Review by bmorris3 — APPROVED** [review]
Thanks @kecnry, looks good 👌🏻

**Pull request merged**
